src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs

```diff
@@ -272,101 +272,127 @@ public static void DeleteFile(string fullPath)
 
         public static void CreateDirectory(string fullPath)
         {
-            // NOTE: This logic is primarily just carried forward from Win32FileSystem.CreateDirectory.
+            // The argument is a full path, which means it is an absolute path that
+            // doesn't contain "//", "/./", and "/../".
+            Debug.Assert(fullPath.Length > 0);
+            Debug.Assert(PathInternal.IsDirectorySeparator(fullPath[0]));
 
-            int length = fullPath.Length;
-
-            // We need to trim the trailing slash or the code will try to create 2 directories of the same name.
-            if (length >= 2 && Path.EndsInDirectorySeparator(fullPath))
+            if (fullPath.Length == 1)
             {
-                length--;
+                // fullPath is '/'.
+                return;
             }
 
-            // For paths that are only // or ///
-            if (length == 2 && PathInternal.IsDirectorySeparator(fullPath[1]))
+            string mkdirPath = fullPath;
+            int result = Interop.Sys.MkDir(mkdirPath, (int)Interop.Sys.Permissions.Mask);
+            if (result == 0)
             {
-                throw new IOException(SR.Format(SR.IO_CannotCreateDirectory, fullPath));
+                // Created directory.
+                return;
             }
 
-            // We can save a bunch of work if the directory we want to create already exists.
-            if (DirectoryExists(fullPath))
+            Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
+            if (errorInfo.Error == Interop.Error.EEXIST)
             {
-                return;
+                // Path already exists. Ensure it's a directory.
+                if (DirectoryExists(fullPath))
+                {
+                    return;
+                }
             }
-
-            // Attempt to figure out which directories don't exist, and only create the ones we need.
-            bool somepathexists = false;
-            List<string> stackDir = new List<string>();
-            int lengthRoot = PathInternal.GetRootLength(fullPath);
-            if (length > lengthRoot)
+            else if (errorInfo.Error == Interop.Error.ENOENT)
             {
-                int i = length - 1;
-                while (i >= lengthRoot && !somepathexists)
+                // Some parts of the path don't exist yet.
+
+                // Try create parents bottom to top and track those that could not
+                // be created due to missing parents. Then create them top to bottom.
+
+                List<string> stackDir = new List<string>();
```

Comment:
Instead of creating a list of strings, we could create a list of integers, where every integer would represent the index for slicing the original string, as they all start from `0`.

Sth like this:

```diff
- mkdirPath = fullPath.Substring(0, i);
- stackDir.Add(mkdirPath);
+ mkdirPath = fullPath.Slice(0, i);
+ stackDir.Add(i);
```

@GrabYourPitchforks is it safe to assume that strings are immutable?

Reply:
I believe we assume strings are immutable throughout. Except possibly some special perf cases where we created the string on the previous line and immediately write to it using a pointer.

Reply:
We eventually need these strings for the `MkDir` PInvoke.

Reply:
It's safe to assume that strings are immutable. __However__, code paths like this have a tendency to get refactored such that their inputs become `ROS<char>` instead of `string`, and at that point all immutability guarantees are void.

That said, what's the worst thing that can happen if the span contents end up changing mid-operation? We accidentally read the wrong directory names, maybe resulting in an i/o error and exception? That's probably ok, not the end of the world. We accidentally read off the end of a buffer and start corrupting memory? That's unacceptable.

So if we pessimistically assume that the contents may be mutable in a future release, and we also think that the consequences of something being mutated out from under us isn't really all that bad, then I wouldn't worry too much about it. :)